### PR TITLE
feat(project-engine-client): typed client wrapper + IMS Bearer auth + retry + spec-correction overlay (LLMO-5461)

### PR DESCRIPTION
**Pull request by aliciadriani**
## What

Adds the **typed Project Engine client wrapper** to `@adobe/spacecat-shared-project-engine-client` — a thin `openapi-fetch` layer over the generated `paths` that owns the base URL, retries, and request authentication — together with a **generation-time spec-correction overlay** that aligns the vendored Semrush swagger with the live API.

Part of **LLMO-5459** (epic) / **LLMO-5461**. Step 2 of the work plan.

## Behaviour

- `createSerenityProjectEngineApiClient(options)` — `openapi-fetch` client over the generated `paths`.
- **Auth:** the caller's **raw IMS JWT** is forwarded verbatim as the `Authorization: Bearer <token>` header — matching the deployed `spacecat-api-service` transport (`src/support/serenity/rest-transport.js`). Semrush accepts the IMS bearer directly; the client mints or exchanges **nothing**. Accepts a static token or a (sync/async) per-request getter, so the header reflects the current caller, not a token captured at construction. Fails fast on an empty/missing token, and propagates a throwing getter (e.g. IMS outage), rather than sending `Bearer undefined`.
- **Base URL:** the client owns the fixed `/enterprise/projects/api` prefix. It normalises the configured base URL to its origin (dropping any path/credentials) and appends the prefix, mirroring `rest-transport.js`. Idempotent for callers already passing the prefix. Invalid or non-http(s) base URLs fail fast at construction.
- **Retry:** `429` retried for **any** method; `5xx` + network errors retried **only for idempotent** methods (GET/HEAD/PUT/DELETE/OPTIONS), so a POST that may have created a resource is never replayed. Bodied requests are `clone()`d per attempt so retries send a fresh, unconsumed body. Backoff is **exponential with equal jitter** (0.5x–1x) to de-correlate concurrent clients, and **honours a `Retry-After` header** (delta-seconds or HTTP-date) when it asks for longer than the backoff — clamped to a 20s ceiling. Defaults: 2 retries, base 200ms. After exhausting retries, returns the last retryable response (e.g. final 503) or rethrows the last network error.

## Spec-correction overlay

The vendored Semrush swagger (`spec/projectengine_swagger_public.yaml`) is kept pristine for re-vendoring. A small OpenAPI Overlay (`spec/overlays/corrections.yaml`), applied to the converted OAS3 artifact at generation time by `scripts/apply-overlay.mjs`, corrects three places where the published spec diverges from the live API:

- **CR1** — adds `GET /v1/ai_models` (the live global model catalog, absent from the upstream swagger).
- **CR2** — removes the required `Auth-Data-Jwt` header from every operation and adds an `imsBearer` security scheme. The live API authenticates on `Authorization: Bearer <IMS>`; `Auth-Data-Jwt` is rejected.
- **CR3** — corrects the `GET /v2/workspaces/{id}/projects` description: it also returns initial, never-yet-published draft projects, not only published ones.

Guard tests pin CR1 and CR2, so a future Semrush spec refresh that silently drops the overlay fails loudly instead of regressing the generated types.

### Auth-header note

An earlier revision forwarded the IMS token as the spec's native `Auth-Data-Jwt` header. That was verified to 401 against the live Semrush endpoint (`adobe-hackathon.semrush.com`), while `Authorization: Bearer <IMS>` returns 200. The client uses Bearer, and the CR2 overlay removes `Auth-Data-Jwt` from the generated TypeScript types and Pydantic models so the contract matches reality.

## Convention port

The scaffold's TypeScript surface (`internal.ts` / `client.ts` / `index.ts` + Vitest tests) is ported one-for-one to **JS + ESM + JSDoc**, **mocha + chai + sinon + c8** — the design does not change, only the surface. A hand-written `src/index.d.ts` re-exports the typed surface plus the generated `paths`/`components`.

## Validation

- `npm test` — 46 passing, **100% statements / branches / functions / lines**
- `npm run lint` — clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/spacecat-shared-project-engine-client/src/internal.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * @param {number} baseDelayMs
+ * @returns {typeof globalThis.fetch}
+ */
+export function createRetryingFetch(baseFetch, maxRetries, baseDelayMs) {
+  return async function retryingFetch(input, init) {
+    const method = methodOf(input, init);
+    // openapi-fetch calls us with a Request object; fetch() consumes its body on use, so a
+    // bare replay throws "Request ... already used". Clone per attempt and never touch the
+    // original, so every retry (incl. a 429 on a bodied POST) sends a fresh, unconsumed body.
+    const forAttempt = () => (input instanceof Request ? input.clone() : input);
```
Comment (aliciadriani) [inline packages/spacecat-shared-project-engine-client/src/internal.js:192]: **Must Fix #1 — resolved.** `openapi-fetch` invokes the injected fetch with a single `Request` object (`fetch(request, requestInitExt)`), and Node's `fetch` consumes the body on first use. Replaying the same `Request` therefore threw `TypeError: Request ... already used` on `POST`+429, and silently skipped the retry (returning the first 503) on `PUT`/`DELETE`+5xx. Now we `clone()` the Request per attempt and never touch the original, so every retry sends a fresh, unconsumed body. Confirmed end-to-end against a real server. Covered by the two new regression tests in `internal.test.js`.

File: packages/spacecat-shared-project-engine-client/src/client.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * Builds the openapi-fetch middleware that forwards the caller's IMS token verbatim as the
+ * `Auth-Data-Jwt` header.
+ * @param {() => string | Promise<string>} getToken
+ * @returns {import('openapi-fetch').Middleware}
+ */
+function authMiddleware(getToken) {
+  return {
+    async onRequest({ request }) {
+      const token = await getToken();
```
Comment (aliciadriani) [inline packages/spacecat-shared-project-engine-client/src/client.js:94]: **Nit (auth guard) — resolved.** Previously a getter resolving to `undefined`/`null`/`''` would `headers.set('Auth-Data-Jwt', undefined)`, sending the literal string `"undefined"` (or an empty header) and earning an opaque 401/403. Now we fail fast with a clear error before the request is sent. Tested in `client.test.js`.

File: packages/spacecat-shared-project-engine-client/test/internal.test.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    base.onCall(1).resolves(resp(201));
+    const res = await createRetryingFetch(base, 2, 0)('https://x', { method: 'POST' });
+    expect(res.status).to.equal(201);
+    expect(base.callCount).to.equal(2);
+  });
+
+  // Regression: openapi-fetch hands us a Request object, and reading its body (as real
+  // fetch does) marks it used — a bare replay would throw "Request ... already used".
+  // These two assert the per-attempt clone actually re-sends the body on retry.
+  it('re-sends a bodied idempotent Request on retry (clones rather than replaying)', async () => {
```
Comment (aliciadriani) [inline packages/spacecat-shared-project-engine-client/test/internal.test.js:92]: **Should Fix #2 — resolved.** The original retry tests passed a string URL + stubbed fetch, so they never exercised the `Request`-with-body path `openapi-fetch` actually produces — which is how 100% line coverage coexisted with a broken retry. These two tests drive `createRetryingFetch` with a real `Request` carrying a body and a stub that reads `req.text()` each call (mirroring real fetch's consumption). They fail against the pre-fix code and pass after the `clone()` fix.

File: packages/spacecat-shared-project-engine-client/test/client.test.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      retryBaseDelayMs: 0,
+      fetch,
+    });
+
+    const { response } = await client.GET('/v1/countries');
+    expect(response.status).to.equal(200);
+    expect(fetch.callCount).to.equal(2);
+  });
+
+  it('fails fast (and never sends the request) when the token resolves to empty', async () => {
```
Comment (aliciadriani) [inline packages/spacecat-shared-project-engine-client/test/client.test.js:109]: **Nit (auth guard) — test.** Asserts the empty-token getter makes the call fail fast *and* that the underlying fetch is never invoked.

File: packages/spacecat-shared-project-engine-client/src/internal.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2026 Adobe. All rights reserved.
```
Comment (aliciadriani) [inline packages/spacecat-shared-project-engine-client/src/internal.js:2]: **Nit (copyright) — resolved.** Bumped the new-file headers from 2025 to 2026 to match the convention for files added to `spacecat-shared` this year. Applied across all 7 new files in the PR.

**Review by aliciadriani — COMMENTED** [review]
Self-review follow-up: pushed `4c7c521` addressing all findings from the review (1 Must Fix, 1 Should Fix, 2 nits). Inline comments below map each fix to its finding. Verified end-to-end against a real `http` server before and after; suite now 27 passing, 100% coverage, lint clean.

**Review by MysticatBot — COMMENTED** [review]
Hey @aliciadriani,

**Verdict: Comment** - clean, well-layered client with solid test coverage; only minor polish items.
_Changes: Adds typed Project Engine API client wrapper (openapi-fetch + Auth-Data-Jwt + retry with idempotency-awareness) with 18 tests at 100% coverage (9 files)._

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: Missing `baseUrl` validation at the factory boundary - passing `undefined` or `''` yields a cryptic error from inside openapi-fetch rather than a fail-fast guard like the auth token has - `packages/spacecat-shared-project-engine-client/src/client.js:52`
- suggestion: Add jitter to retry backoff (`delay * (0.5 + Math.random() * 0.5)`) to prevent thundering-herd alignment when multiple Lambda instances hit the same 429/503 - `packages/spacecat-shared-project-engine-client/src/internal.js:92`
- suggestion: Respect `Retry-After` header on 429 responses - the current fixed exponential backoff may be shorter than the server's requested wait, leading to repeated 429s under aggressive rate limiting
- nit: Missing test for a throwing token getter (e.g. IMS outage rejecting the promise) - distinct from the empty-string path and a scenario real callers will hit - `packages/spacecat-shared-project-engine-client/test/client.test.js`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 52s | <b>Cost:</b> $3.86 | <b>Commit:</b> <code>b052eac652c59ab4e454ae3ee0b937396d9d5536</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @aliciadriani,

**Verdict: Approve** - clean, well-layered client with all prior findings addressed; only minor polish items remain.
_Changes: Adds typed Project Engine API client wrapper (openapi-fetch + IMS Bearer auth + idempotency-aware retry with jitter and Retry-After) with 40 tests at 100% coverage (10 files)._

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: No guard on negative `maxRetries` - passing a negative value causes the retry loop to skip, leaving both `lastResponse` and `lastError` undefined, which throws `undefined` (confusing). A `Math.max(0, maxRetries)` floor would fail more clearly. - `packages/spacecat-shared-project-engine-client/src/internal.js:100`
- nit: `resolveBaseUrl` does not restrict the protocol to `http:`/`https:` - a `file://` or `ftp://` URL parses successfully but fails later at `fetch()` with a less actionable error. Low risk since base URL comes from service config, not user input. - `packages/spacecat-shared-project-engine-client/src/client.js:62`
- nit: Test files use `sinon.stub()`/`sinon.restore()` directly rather than the `sinon.createSandbox()` pattern documented in CLAUDE.md under "Test Conventions". Functionally equivalent but diverges from the convention other packages follow. - `packages/spacecat-shared-project-engine-client/test/internal.test.js`
- suggestion: Add edge-case tests for `parseRetryAfterMs` with negative delta-seconds (`"-5"` floors to 0) and a `maxRetries=0` boundary test (fetch called exactly once, no retry) to lock in documented behavior at the edges. - `packages/spacecat-shared-project-engine-client/test/internal.test.js`
</details>

## Previously flagged, now resolved
- baseUrl validation now fails fast at construction with actionable error message
- Retry backoff now applies equal jitter (0.5x-1x) to de-correlate concurrent clients
- Retry-After header honoured (delta-seconds and HTTP-date), clamped to 20s ceiling
- Throwing token getter test added (IMS outage scenario)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 5s | <b>Cost:</b> $4.65 | <b>Commit:</b> <code>b14096e560345c5437efcfc78a0855ba6e345733</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — COMMENTED** [review]
## Review — PR #1661

Re-review (large update, 52 commits since last approved review). Full review performed. All prior findings have been addressed.

**Verdict: Approve with suggestions** — solid implementation, well-tested (100% coverage), minimal dependency surface. The issues below are improvements, not blockers.

---

### Important

**1. 429 retry on non-idempotent POST risks duplicate creation**
`src/internal.js:83-86` — `isRetryableStatus` returns true for 429 regardless of HTTP method. The inline comment assumes Semrush's 429 is always pre-processing, but this is a fragile upstream contract. Multiple POST endpoints in the spec perform create operations (bulk-create-projects, create-brand-urls, create-prompt).
**Suggestion:** Add an explicit inline comment documenting the assumption and the upstream behavior that validates it. If Semrush's rate-limiting architecture ever changes, POST retries become unsafe. Consider an idempotency-key header if the API supports one.

**2. No observability hook for retry loops**
`src/internal.js` — Retry loops can silently delay responses by up to 60s (3 attempts × 20s cap) with no log, metric, or callback. An operator cannot distinguish "slow upstream" from "stuck in retry backoff" without code-level debugging.
**Suggestion:** Accept an optional `onRetry(attempt, delayMs, status)` callback (or emit a structured log) so consumers can wire observability without modifying internal.js.

**3. README states client is "later PR" — now stale**
`README.md:6` and the intro paragraph both describe the client wrapper as landing in a "follow-up PR." This PR delivers that client. The README will actively mislead readers about what the package contains.
**Suggestion:** Update the README to reflect the current state — the client wrapper is implemented. A one-paragraph description of the client's auth + retry behavior would be appropriate.

**4. `apply-overlay.mjs` (189 lines) has zero unit tests**
`scripts/apply-overlay.mjs` — The overlay script implements JSONPath-like traversal with `remove`/`add`/`replace` operations. Its `remove` operation with 0 matches silently succeeds, which could mask a regression if the vendored spec changes shape and a correction no longer hits its target.
**Suggestion:** Add basic tests for the overlay operations — particularly the 0-match case for `remove`, which should at minimum warn or throw rather than silently succeed.

**5. No test for auth-token reuse across retries**
`test/client.test.js` — Tests verify auth is attached per-request, but no test verifies that `getToken()` is re-invoked on each retry attempt (vs. caching a stale token from the first failed attempt). If a token expires mid-retry-loop, the retry would repeat with the same expired token.
**Suggestion:** Add a test where the token getter returns different values on successive calls, verify retries use fresh tokens.

**6. `resolveBaseUrl` credential-stripping not tested**
`src/client.js:63-65` — The function strips `username`/`password` from URLs containing embedded credentials, but no test exercises this path. This is a security-relevant behavior that should have explicit coverage.
**Suggestion:** Add a test with `https://user:pass@host/path` input, assert the output URL has no credentials.

---

### Minor

- **`init` param alongside cloned Request** (`src/internal.js:97`) — Passing both a cloned `Request` and an `init` object to `fetch(req, init)` can produce surprising precedence conflicts depending on the runtime. Consider spreading headers into the cloned request directly.
- **Custom JSONPath engine maintenance** (`scripts/apply-overlay.mjs`) — The hand-rolled path traversal works today but will need upkeep as overlay corrections grow. Consider migrating to a lightweight JSONPath library if corrections exceed ~5.
- **`afterEach` at module scope** (`test/client.test.js`) — Sinon `afterEach` restore at the top module scope instead of inside a `describe` block risks unintended interaction with other test files in the same process.
- **CR3 description is 330+ chars on one line** (`spec/overlays/corrections.yaml`) — Long single-line strings in YAML are hard to review in diffs. Consider a YAML block scalar (`|`).
- **`index.d.ts` references "CR2" jargon** — The comment in the public type declaration references internal correction numbering that external consumers won't understand.
- **Functional pattern vs. documented class+factory** — The `createSerenityProjectEngineApiClient` function diverges from the `CLAUDE.md`-documented "class + factory" API client pattern. The functional approach is reasonable for a thin wrapper; consider adding a one-line note in the package README explaining the choice.
- **`toTokenGetter` passes through non-string/non-function values** (`src/internal.js:167-168`) — If `authToken` is accidentally `null`, `0`, or an object, it flows through to the `Authorization` header. A type guard at construction time would surface misconfiguration earlier.

---

### Positive

- Minimal dependency surface (single prod dep, pinned, MIT)
- Defensive URL handling — protocol restriction, credential stripping, path normalization
- Well-designed retry: exponential backoff with jitter, Retry-After support, idempotency-aware for 5xx, hard ceiling at 20s
- 100% test coverage with behavior-focused assertions
- Clean overlay system that patches vendored spec at generate-time without modifying the source file

---

*Re-review: all findings from the previous approved review have been addressed. This review covers the full 52-commit delta.*

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 4m 9s | <b>Cost:</b> $7.53 | <b>Commit:</b> <code>2dab77d6bed663aa751fd1c11674ec457e7ab522</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: packages/spacecat-shared-project-engine-client/src/internal.js [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+ * observability callback can never break the retry loop or the request.
+ * @param {OnRetry} [onRetry]
+ * @param {object} info
+ */
+function notifyRetry(onRetry, info) {
+  if (!onRetry) {
+    return;
+  }
+  try {
+    onRetry(info);
```
Comment (MysticatBot) [inline packages/spacecat-shared-project-engine-client/src/internal.js]: issue (blocking): The `try { onRetry(info); } catch {}` pattern only catches synchronous throws. If a consumer passes an `async` function (which TypeScript allows, since `() => void` accepts async callbacks), the returned Promise is silently dropped. If that Promise rejects, it becomes an unhandled promise rejection - which crashes the Node process in Node 18+ by default.

The PR documentation says "A throwing hook is swallowed and never affects the request" but this guarantee only holds for synchronous throws. Since observability hooks naturally tend to be async (metrics endpoints, structured loggers), this is a realistic footgun.

Fix (one line, inside `notifyRetry`):
```js
try {
  const result = onRetry(info);
  if (result && typeof result.catch === 'function') {
    result.catch(() => {});
  }
} catch {
  // best-effort
}
```

This defensively sinks any returned promise without changing the synchronous contract or the timing (the hook still fires before `sleep`, and awaiting is not needed since the callback is best-effort).

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @aliciadriani,

**Verdict: Request changes** - one subtle safety gap in the new onRetry hook; otherwise clean and ready.
_Changes: Addresses all prior review findings (onRetry observability hook, overlay tests, credential-stripping test, type guard, README update) across 10 files in the project-engine-client package (5 commits)._

## Must fix before merge

1. [Important] `notifyRetry` does not catch async rejections - an async `onRetry` callback will crash the process via unhandled promise rejection - `packages/spacecat-shared-project-engine-client/src/internal.js:136` (details inline)

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- nit: `notifyRetry` empty catch produces zero diagnostic output when the hook throws - a broken observer is invisible to operators. Consider a single `console.warn` on first failure or document explicitly that no signal is emitted. - `packages/spacecat-shared-project-engine-client/src/internal.js:137`
- suggestion: `applyOverlay()` uses bare `console.log`/`console.warn` inside the exported function. Callers importing it programmatically (rather than via CLI `main()`) will get uncontrolled stdout. The CLI entry point is the right place for user-facing output; the function itself could return structured results instead. - `packages/spacecat-shared-project-engine-client/scripts/apply-overlay.mjs:170`
- suggestion: Python models (`python/serenity_project_engine/`) are committed with no `pyproject.toml`, no consumer, and a `pip install` requirement for generation. Worth addressing before the package goes public or before a second consumer arrives.
</details>

## Previously flagged, now resolved
- README updated to reflect current state (client implemented, overlay documented)
- `apply-overlay.mjs` now has 158 lines of unit tests covering all branches
- `onRetry` observability hook added with tests
- Auth-token reuse across retries tested
- Credential-stripping behavior tested
- `toTokenGetter` type guard rejects non-string/non-function at construction
- CR3 description reformatted as YAML folded scalar
- `index.d.ts` "CR2" jargon removed

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 4m 32s | <b>Cost:</b> $8.42 | <b>Commit:</b> <code>a32500ad6a8fda523f5af42a89a69d37c98390a7</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @aliciadriani,

**Verdict: Approve** - prior blocking finding addressed cleanly; only minor polish items remain.
_Changes: Fixes async onRetry rejection handling, refactors applyOverlay to return structured results instead of logging, updates types and tests (7 files)._

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: `@callback OnRetry` JSDoc declares `@returns {void}` but `index.d.ts` now correctly types the return as `void | Promise<void>` - the two are out of sync for IDE tooling consumers of the JSDoc - `packages/spacecat-shared-project-engine-client/src/internal.js:158`
- suggestion: The async-rejection test proves "the request succeeds" rather than "no rejection escaped" - it relies on Node 18+ crashing the process on unhandled rejections rather than asserting explicitly. Consider a scoped `process.on('unhandledRejection', fail)` listener for a positive assertion. Low priority since the test runner enforces this at the process level. - `packages/spacecat-shared-project-engine-client/test/internal.test.js:210`
</details>

## Previously flagged, now resolved
- Async onRetry rejection now sunk via duck-typed `.catch()` on any thenable return value
- Overlay tests cleaned up: sinon stubs removed after applyOverlay refactored to pure function returning structured results

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 42s | <b>Cost:</b> $4.55 | <b>Commit:</b> <code>9023e1dd6b9f8a00b2e37313967d6a6b139f9f16</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @aliciadriani,

**Verdict: Approve** - prior blocking finding cleanly addressed; only minor polish items remain.
_Changes: Addresses async onRetry rejection handling (the prior blocking finding) and refactors overlay script to return structured results instead of logging directly (7 files)._

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: Duck-typing `.catch` is looser than the standard thenable check - `typeof result?.then === 'function'` is the ecosystem idiom and aligns better with the `void | Promise<void>` return type. Practical risk is negligible since `onRetry` is caller-controlled. - `packages/spacecat-shared-project-engine-client/src/internal.js:141`
- suggestion: The async onRetry test implicitly relies on mocha surfacing unhandled rejections as failures. An explicit scoped `process.on('unhandledRejection')` listener would make the test intent more visible and guard against future runners that swallow them. Strictly optional. - `packages/spacecat-shared-project-engine-client/test/internal.test.js:210`
</details>

## Previously flagged, now resolved
- Async onRetry rejection now sunk via duck-typed `.catch()` with dedicated test
- `applyOverlay()` returns structured `{ total, results }` with no console output; CLI `main()` handles logging
- JSDoc expanded to document the deliberate silent-failure design choice

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 24s | <b>Cost:</b> $4.60 | <b>Commit:</b> <code>dc6c3269cc05b558e69c45236b543086591be148</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by rainer-friederich — APPROVED** [review]
LGTM

**Review by MysticatBot — APPROVED** [review]
Hey @aliciadriani,

**Verdict: Approve** - prior nits addressed cleanly; no new issues.
_Changes: Syncs OnRetry JSDoc return type with index.d.ts and adds positive unhandledRejection assertion to the async hook test (2 files)._

<details>
<summary>Non-blocking (1): minor issues and suggestions</summary>

- nit: Test comment says "macrotask turn" but `setImmediate` is technically a check-phase callback, not a macrotask - "event-loop turn" would be more precise - `packages/spacecat-shared-project-engine-client/test/internal.test.js:217`
</details>

## Previously flagged, now resolved
- OnRetry `@callback` JSDoc `@returns` now `void | Promise<void>`, matching `index.d.ts`
- Async-rejection test uses scoped `process.on('unhandledRejection')` listener for positive assertion with proper cleanup

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 5m 29s | <b>Cost:</b> $1.78 | <b>Commit:</b> <code>1e4411b53b6e64190554f59177d8024c5154441a</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @aliciadriani,

**Verdict: Approve** - clean, focused security hardening with correct test coverage.
_Changes: Hardens `deepMerge` against prototype pollution by skipping `__proto__`, `constructor`, and `prototype` keys, with a regression test exercising the real attack vector via `JSON.parse` (2 files)._

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 2s | <b>Cost:</b> $2.16 | <b>Commit:</b> <code>1c7411792bee73a96cac2582afd348ecbc18e7a7</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by rainer-friederich — APPROVED** [review]
LGTM

**Review by rainer-friederich — APPROVED** [review]
LGTM
